src/text-buffer.coffee

```diff
@@ -1138,7 +1172,22 @@ class TextBuffer
       else
         iterator = new MatchIterator.ForwardsMultiLine(this, regex, range)
 
-    iterator.iterate(callback, global)
+    callback_with_context = (argument) =>
+      argument.linesBefore = []
+      for i in [0...(options.lineCountBefore or 0)]
+        row = argument.range.start.row - i - 1
+        break if row < 0
+        argument.linesBefore.unshift(@lineForRow(row))
+
+      argument.linesAfter = []
+      for i in [0...(options.lineCountAfter or 0)]
+        row = argument.range.start.row + i + 1
+        break if row >= @getLineCount()
+        argument.linesAfter.push(@lineForRow(row))
+
+      callback(argument)
```

Comment:
Since `scanInRange` is called a lot, I'd like to avoid creating this wrapper closure. Could you move this functionality to the `SingleLineSearchCallbackArgument` and `MultiLineSearchCallbackArgument` classes?

I think that `iterator.iterate` could take `leadingContextLineCount` and `trailingContextLineCount` as two new parameters, and could pass them to the callback arguments' constructors.